nomad/state/events_state_store.go

```diff
@@ -0,0 +1,47 @@
+package state
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// addNodeEvent is a function which wraps upsertNodeEvent
+func (s *StateStore) AddNodeEvent(index uint64, nodeID string, event *structs.NodeEvent) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	return s.upsertNodeEvent(index, nodeID, event, txn)
+}
+
+// upsertNodeEvent upserts a node event for a respective node. It also maintains
+// that only 10 node events are ever stored simultaneously, deleting older
+// events once this bound has been reached.
+func (s *StateStore) upsertNodeEvent(index uint64, nodeID string, event *structs.NodeEvent, txn *memdb.Txn) error {
+
+	ws := memdb.NewWatchSet()
```

Comment:
Look how `UpdateNodeStatus` is done. This method needs to:
1. access the node directly, not via NodeByID
2. Copy the Node and reinsert it. (treat objects as immutable). This data structure allows multiple readers to be accessing it so you are modifying a consistent view if you don't copy and re-insert.
3. Update the Node's Modify index otherwise blocking queries will not work.